File: backend/app/schemas/kg_ledger.py
"""
Schemas KgLedger (SAC E2, v0.5 §12.1.1, plan-sac-e2 D1/D6/D14/D15/D16).

El Literal de creacion manual solo permite 'manual_adjustment' (D1): los
source_types de negocio (postconsumo_receipt, drosses_receipt) los emite SOLO
el servicio de inbound; los 9 restantes de §4.3 llegan con E3/E4.
"""
from datetime import datetime, timezone
from decimal import Decimal
from typing import Literal, Optional
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field, field_validator

from app.utils.dates import BusinessDate

KgAccountType = Literal[
    "willard_baterias", "willard_drosses", "intersede", "intra_horno", "crisol"
]


class KgLedgerAccountCreate(BaseModel):
    model_config = ConfigDict(extra="forbid")

    code: str = Field(..., min_length=1, max_length=32)
    display_name: str = Field(..., min_length=1, max_length=120)
    account_type: KgAccountType
    warehouse_id: Optional[UUID] = None
    third_party_id: Optional[UUID] = None
    tolerance_kg: Optional[Decimal] = Field(
        None, ge=0, description="Tolerancia del cuadre semanal (± kg); NULL = sin alerta"
    )


class KgLedgerAccountUpdate(BaseModel):
    """Solo metadata — account_type y FKs son inmutables post-creacion (§12.1.1)."""
    model_config = ConfigDict(extra="forbid")

    display_name: Optional[str] = Field(None, min_length=1, max_length=120)
    tolerance_kg: Optional[Decimal] = Field(None, ge=0)
    is_active: Optional[bool] = None


class KgLedgerAccountResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    code: str
    display_name: str
    account_type: str
    warehouse_id: Optional[UUID] = None
    warehouse_name: Optional[str] = None
    third_party_id: Optional[UUID] = None
    third_party_name: Optional[str] = None
    tolerance_kg: Optional[Decimal] = None
    current_balance_kg: Decimal = Decimal("0")
    is_active: bool
    created_at: datetime


class KgLedgerMovementManualCreate(BaseModel):
    """Movimiento manual auditado (source_type='manual_adjustment', §4.3)."""
    model_config = ConfigDict(extra="forbid")

    account_id: UUID
    delta_kg: Decimal
    transaction_date: BusinessDate
    description: str = Field(..., min_length=1, max_length=300)
    reason: str = Field(..., min_length=1, max_length=180, description="Motivo obligatorio")

    @field_validator("delta_kg")
    @classmethod
    def delta_nonzero(cls, v: Decimal) -> Decimal:
        if v == 0:
            raise ValueError("delta_kg no puede ser 0 — los movimientos nulos se anulan, no se insertan")
        return v

    @field_validator("transaction_date")
    @classmethod
    def not_future(cls, v: datetime) -> datetime:
        # Canon §11.1.2 L1917 + filosofia anti back-dating #62 (D15)
        if v.date() > datetime.now(timezone.utc).date():
            raise ValueError("transaction_date no puede ser futura")
        return v


class KgLedgerMovementResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    account_id: UUID
    delta_kg: Decimal
    transaction_date: datetime
    description: Optional[str] = None
    source_type: str
    source_id: Optional[UUID] = None
    inventory_movement_id: Optional[UUID] = None
    conversion_formula_snapshot: Optional[dict] = None
    status: str
    annulled_reason: Optional[str] = None
    annulled_at: Optional[datetime] = None
    created_at: datetime


class KgLedgerStatementRow(KgLedgerMovementResponse):
    balance_after_kg: Decimal


class KgLedgerStatementResponse(BaseModel):
    account: KgLedgerAccountResponse
    opening_balance_kg: Decimal
    movements: list[KgLedgerStatementRow]
    current_balance_kg: Decimal


class KgLedgerSummaryAccount(BaseModel):
    account_id: UUID
    code: str
    display_name: str
    account_type: str
    warehouse_id: Optional[UUID] = None
    warehouse_name: Optional[str] = None
    balance_kg: Decimal
    tolerance_kg: Optional[Decimal] = None
    last_movement_at: Optional[datetime] = None
    is_active: bool


class KgLedgerSummaryResponse(BaseModel):
    accounts: list[KgLedgerSummaryAccount]
    total_willard_kg: Decimal
    total_intersede_kg: Decimal
    total_intra_horno_kg: Decimal
    total_crisol_kg: Decimal


class KgLedgerAnnulRequest(BaseModel):
    reason: str = Field(..., min_length=1, max_length=500)

File: frontend/src/types/kg-ledger.ts
// Tipos KgLedger — cuentas en kilogramos de plomo (SAC E2)
// Espejo de backend/app/schemas/kg_ledger.py

export type KgAccountType =
  | "willard_baterias"
  | "willard_drosses"
  | "intersede"
  | "intra_horno"
  | "crisol";

export type KgMovementStatus = "confirmed" | "annulled";

export const KG_ACCOUNT_TYPE_LABELS: Record<KgAccountType, string> = {
  willard_baterias: "Willard Baterías",
  willard_drosses: "Willard Drosses",
  intersede: "Intersede",
  intra_horno: "Horno",
  crisol: "Crisol",
};

// Coherencia tipo↔FKs — espejo de services/kg_ledger.py (el backend valida con 422)
export const KG_WILLARD_TYPES: KgAccountType[] = ["willard_baterias", "willard_drosses"];
export const KG_WAREHOUSE_REQUIRED_TYPES: KgAccountType[] = [
  "willard_baterias",
  "intra_horno",
  "crisol",
];

export const KG_SOURCE_TYPE_LABELS: Record<string, string> = {
  manual_adjustment: "Ajuste manual",
  postconsumo_receipt: "Recepción postconsumo",
  drosses_receipt: "Recepción drosses",
  migration_initial_load: "Carga inicial migración",
};

export interface KgLedgerAccountResponse {
  id: string;
  code: string;
  display_name: string;
  account_type: KgAccountType;
  warehouse_id: string | null;
  warehouse_name: string | null;
  third_party_id: string | null;
  third_party_name: string | null;
  tolerance_kg: number | null;
  current_balance_kg: number;
  is_active: boolean;
  created_at: string;
}

export interface KgLedgerAccountCreate {
  code: string;
  display_name: string;
  account_type: KgAccountType;
  warehouse_id?: string | null;
  third_party_id?: string | null;
  tolerance_kg?: number | null;
}

export interface KgLedgerAccountUpdate {
  display_name?: string;
  tolerance_kg?: number;
  is_active?: boolean;
}

export interface KgLedgerMovementManualCreate {
  account_id: string;
  delta_kg: number;
  transaction_date: string;
  description: string;
  reason: string;
}

export interface KgLedgerMovementResponse {
  id: string;
  account_id: string;
  delta_kg: number;
  transaction_date: string;
  description: string | null;
  source_type: string;
  source_id: string | null;
  inventory_movement_id: string | null;
  conversion_formula_snapshot: Record<string, unknown> | null;
  status: KgMovementStatus;
  annulled_reason: string | null;
  annulled_at: string | null;
  created_at: string;
}

export interface KgLedgerStatementRow extends KgLedgerMovementResponse {
  balance_after_kg: number;
}

export interface KgLedgerStatementResponse {
  account: KgLedgerAccountResponse;
  opening_balance_kg: number;
  movements: KgLedgerStatementRow[];
  current_balance_kg: number;
}

export interface KgLedgerSummaryAccount {
  account_id: string;
  code: string;
  display_name: string;
  account_type: KgAccountType;
  warehouse_id: string | null;
  warehouse_name: string | null;
  balance_kg: number;
  tolerance_kg: number | null;
  last_movement_at: string | null;
  is_active: boolean;
}

export interface KgLedgerSummaryResponse {
  accounts: KgLedgerSummaryAccount[];
  total_willard_kg: number;
  total_intersede_kg: number;
  total_intra_horno_kg: number;
  total_crisol_kg: number;
}
